### PR TITLE
MooseVariable -> MooseObject

### PR DESCRIPTION
**Pull request by lindsayad**
Will allow:

- MooseVariable to have restartable data
- consistent interface/behavior with other MooseObjects
- cleaner code-base

### CODE REVIEW

**Comment by lindsayad** [comment]
@aeslaughter What would be your recommendation for resolving the Documentation failure? I'm thinking about making the `AddElementalFieldAction` triggered by indicator and marker syntax not inherit from `AddVariableAction` since I already have to do some workarounds to prevent the `type` of the Indicator or Marker from being used in the `AddElementalFieldAction`.

But that still leaves me with the fact that both `AddVariableAction` and `AddAuxVariableAction` both build MooseVariables, and I don't want to change that. Should I write documentation pages for the different MooseVariable types in `/Variables` and then just sym-link to those files from `/AuxVariables`?

**Comment by lindsayad** [comment]
moose test timeout will be fixed by https://github.com/libMesh/libmesh/pull/2190

**Comment by aeslaughter** [comment]
@lindsayad I would just symlink to the pages that are the same.

**Comment by lindsayad** [comment]
> I would just symlink to the pages that are the same.

I like it!

File: framework/doc/content/source/variables/MooseVariableBase.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+`SystemBase`. Two classes inherit directly from `MooseVariableBase`:
+`MooseVariableFEBase` and `MooseVariableScalar`. `MooseVariableScalar` represents a
+Moose variable that is constant over the spatial domain. It has a number of
+degrees of freedom equal to the order of the variable, e.g. the following
+variable block would declare a `MooseVariableScalar` with two associated degrees
+of freedom:
+
+```
+[AuxVariables]
+  [./aux_scalar]
```
Comment (aeslaughter) [inline framework/doc/content/source/variables/MooseVariableBase.md]:
```suggestion
  [aux_scalar]
```

File: framework/doc/content/source/variables/MooseVariableBase.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    order = SECOND
+    family = SCALAR
+  [../]
```
Comment (aeslaughter) [inline framework/doc/content/source/variables/MooseVariableBase.md]:
```suggestion
  []
```

**Comment by lindsayad** [comment]
Finally no app failures. Woo!

**Review by aeslaughter — DISMISSED** [review]

File: framework/include/utils/InputParameters.h [hunk 3]
```diff
@@ -916,6 +931,8 @@ class InputParameters : public Parameters
 
   // These are the only objects allowed to _create_ InputParameters
   friend InputParameters emptyInputParameters();
+  // friend InputParameters validParams<MooseApp>();
+  // friend class MooseVariableBase; //TODO:MooseVariableToMooseObject
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/utils/InputParameters.h]: they were left over from you haha. I'll remove them

**Review by aeslaughter — APPROVED** [review]

**Comment by gambka** [comment]
@lindsayad This causes a large amount of failures in modules and other applications when executed with the --error-deprecated parameter passed to ./run_tests related to an API change in the adding of Variables and AuxVariables.

I particularly dislike having the deprecated syntax in the code as new users may take that and continue to use it.  Should moose, the modules, and applications for which we test regularly pass the error-deprecated when making these sort of fundamental API changes? Then the deprecation error is there for untracked apps/inputs to give time for their transition?

**Comment by lindsayad** [comment]
> This causes a large amount of failures in modules and other applications when executed with the --error-deprecated parameter passed to ./run_tests related to an API change in the adding of Variables and AuxVariables.

It seems that the most deprecation messages are about the switch from MeshModifiers from MeshGenerators...but yes I definitely see some from this PR as I would expect.

>  Should moose, the modules, and applications for which we test regularly pass the error-deprecated when making these sort of fundamental API changes? Then the deprecation error is there for untracked apps/inputs to give time for their transition?

We do deprecation testing all over CIVET, like [here](https://moosebuild.inl.gov/job/199724/) and [here](https://civet.inl.gov/job/382429/), which I'm sure you're aware of because that is probably what brought this to your attention. So what change are you proposing? That we make those oranges boxes be red? Is the hope then that the MOOSE developer be the one who goes through and updates all the applications to use the new API?

**Comment by gambka** [comment]
I found it out because I was waiting for the MeshGenerators merge to go in to put a PR to fix the majority of deprecation errors (there is some we can't fix due to another issue #13841). It looks like more are coming from the MeshModifiers/nesting of MeshGenerators because the max_failures is limited to 50.  If that is increased, the majority of failures come from this because we have a lot of tests that use the actions (which appears to be what this API change is for).  

I think as a group we should try to keep the deprecated boxes from ever being orange (or as infrequently as possible). As at some point someone will take the deprecated approach as an example and use and there will be more work when we remove the support for the deprecated approach.  Sure CIVET does testing but if I wasn't waiting for the MeshGenerators I wouldn't have noticed these other failures.  Some people do not update their code that frequently to see the deprecation messages either. Perhaps an email to the mailing list when such a fundamental change to the API is changed? I am fine with making the necessary API changes to the downstream apps I use but as a group we should try to make all our source code and tests use non-deprecated versions of the code as soon as possible.

**Comment by lindsayad** [comment]
> I think as a group we should try to keep the deprecated boxes from ever being orange (or as infrequently as possible).

I agree 100%.

> As at some point someone will take the deprecated approach as an example and use and there will be more work when we remove the support for the deprecated approach.

Absolutely a concern, although I know we MOOSE developers are actually notoriously bad at ever removing the deprecated API.

> Sure CIVET does testing but if I wasn't waiting for the MeshGenerators I wouldn't have noticed these other failures. Some people do not update their code that frequently to see the deprecation messages either. Perhaps an email to the mailing list when such a fundamental change to the API is changed?

I know we regularly check the CIVET pages and it would be nice if our application developers did as well, but an email could be a good thing. Most of our users will have no idea what we are talking about, but that's probably ok.

> I am fine with making the necessary API changes to the downstream apps

I really appreciate it. We do check the civet pages regularly, and certainly updating dowstream apps to my new API is on my TO-DO list, but if an application developer gets to it before me, it really makes me happy.

**Pull request merged**